File: BinaryTree_from_inorder_postorder.py
// Time Complexity : O(N)
// Space Complexity : O(H)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no

// Your code here along with comments explaining your approach: using recursion


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> TreeNode:
        
        if inorder == None or len(inorder) == 0:
            return None
        
        ind = inorder.index(postorder[-1])
        root = TreeNode(postorder[-1])
        
        root.right = self.buildTree(inorder[ind+1:], postorder[ind:-1])
        root.left = self.buildTree(inorder[0:ind], postorder[0:ind])
        
        return root

File: BinaryTreeFromInorderPostorder.java
// Time Complexity: O(n)
// Space Complexity: O(n)

import java.util.Arrays;

public class BinaryTreeFromInorderPostorder {
    public TreeNode buildTree(int[] inorder, int[] postorder) {
        if (postorder == null || inorder == null || postorder.length == 0 || inorder.length == 0) {
            return null;
        }

        TreeNode root = new TreeNode(postorder[postorder.length - 1]);

        if (postorder.length == 1) {
            return root;
        }

        int index = -1;

        for (int i = 0; i < postorder.length; i++) {
            if (inorder[i] == root.val) {
                index = i;
                break;
            }
        }

        int[] postleft = Arrays.copyOfRange(postorder, 0, index);
        int[] inleft = Arrays.copyOfRange(inorder, 0, index);
        int[] postright = Arrays.copyOfRange(postorder, index, postorder.length - 1);
        int[] inright = Arrays.copyOfRange(inorder, index + 1, inorder.length);

        root.left = buildTree(inleft, postleft);
        root.right = buildTree(inright, postright);

        return root;
    }
}
